File: twitch_hookup.js
import tmi from 'tmi.js';
import fs from 'fs';
import path from 'path';
import readline from 'readline';
import { askQuestion } from './index.js'; // Import bot logic

const ACTIVE_CHANNEL = 'peepingotter';
let chatters = [];

// Array to store chat messages
let conversationData = [];

// Twitch chat configuration
const client = new tmi.Client({
    options: { debug: true },
    connection: {
        reconnect: true,
        secure: true,
    },
    identity: {
        username: 'peepingnami', // Replace with your Twitch bot username
        password: 'oauth:51ioccg8wrmo5m9dpb9kc039h5i6wc', // Replace with your OAuth token
    },
    channels: [ACTIVE_CHANNEL], // Channel to join
});

// Connect to Twitch
client.connect().catch(console.error);

// Set up console input
const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

// Start the input loop once connected
client.on('connected', () => {
    console.log('Connected to Twitch chat.');
});

// Listen for chat messages
client.on('message', async (channel, tags, message, self) => {
    if (self) return;

    const username = tags['display-name'];

    // Add the username to the array if it's not already there
    if (!chatters.includes(username)) {
        chatters.push(username);
        console.log(`New user added: ${username}`);
    }
    
    const chatEntry = {
        role: "user",
        content: `${tags['display-name']}: ${message}`
    };

    appendMessage(chatEntry);

    // Send the message to the bot and get a response
    try {
        const botReply = await askQuestion(message, tags['display-name']);
        conversationData.push(chatEntry);

        if (botReply) {
            const botReplyEntry = {
                role: "assistant",
                content: `${botReply}`
            };

            conversationData.push(botReplyEntry);
            appendMessage(botReplyEntry);            
            client.say(channel, `@${tags['display-name']} ${botReply}`); // Reply to the user in Twitch chat
        } 
    } catch (error) {
        console.error('Error processing Twitch message:', error);
    }
});

// Gracefully handle disconnect
process.on('SIGINT', () => {
    console.log('Disconnecting from Twitch...');
    saveConversations(); // Save the full conversation before exit   
});

// Save conversation to JSON file
function saveConversations() {
    console.log(chatters); // This will log all chatters who have chatted
    console.log('Reached save conversation function');
    
    // Iterate through each username in the chatters array
    for (const username of chatters) {
        console.log('Found username:', username);

        // Define the file path for the user profile conversation JSON
        const filename = path.join('/conversations/user_profiles/', `conversation_${username}.json`);

        // Check if the file exists
        if (fs.existsSync(filename)) {
            console.log('Found username\'s file:', filename);
            // You can proceed with further processing here (e.g., reading the file, summarizing it, etc.)
        } else {
            console.log('No file found for username:', username);
            console.log(filename)
        }
    }

    // Disconnect the Twitch client and exit the process after checking all files
    client.disconnect();
    process.exit();
}

// Function to append a single message to the JSON file

export function appendMessage(chatEntry) {
    // Directory to save the conversation file
    const dir = './conversations/chat_logs';
    if (!fs.existsSync(dir)) {
        fs.mkdirSync(dir);
    }

    const filename = `twitchchatconvervation_${ACTIVE_CHANNEL}.json`;
    const filePath = path.join(dir, filename);

    try {
        if (!fs.existsSync(filePath)) {
            // If the file doesn't exist, create it with an empty array
            fs.writeFileSync(filePath, JSON.stringify([chatEntry], null, 2), "utf8");
        } else {
            // Read the current file contents
            const fileData = fs.readFileSync(filePath, "utf8");
            const jsonData = JSON.parse(fileData);

            // Append the new chat entry
            jsonData.push(chatEntry);

            // If there are more than 80 entries, remove the earliest 40
            if (jsonData.length > 80) {
                jsonData.splice(0, 40);  // Remove the first 40 entries
            }

            // Write back the updated data to the file
            fs.writeFileSync(filePath, JSON.stringify(jsonData, null, 2), "utf8");
        }
    } catch (error) {
        console.error('Error appending message:', error);
    }
}

File: twitch_hookup.py
import asyncio
import json
import os
from twitchio.ext import commands

ACTIVE_CHANNEL = 'peepingotter'
chatters = []
conversation_data = []


# Bot Configuration
class TwitchBot(commands.Bot):
    def __init__(self):
        super().__init__(
            token="oauth:51ioccg8wrmo5m9dpb9kc039h5i6wc",  # Replace with your OAuth token
            prefix="!",
            initial_channels=[ACTIVE_CHANNEL],
        )

    async def event_ready(self):
        print(f"Logged in as {self.nick}")
        print(f"Connected to channel: {ACTIVE_CHANNEL}")

    async def event_message(self, message):
        # Ignore the bot's own messages
        if message.echo:
            return

        username = message.author.name
        content = message.content

        # Add the user to the chatters list if not already present
        if username not in chatters:
            chatters.append(username)
            print(f"New user added: {username}")

        # Log the message
        chat_entry = {
            "role": "user",
            "content": f"{username}: {content}",
        }
        append_message(chat_entry)

        # Send the message to the bot and get a response
        try:
            bot_reply = await ask_question(content, username)
            conversation_data.append(chat_entry)

            if bot_reply:
                bot_reply_entry = {
                    "role": "assistant",
                    "content": bot_reply,
                }
                conversation_data.append(bot_reply_entry)
                append_message(bot_reply_entry)

                # Reply to the user in Twitch chat
                await message.channel.send(f"@{username} {bot_reply}")
        except Exception as e:
            print(f"Error processing message from {username}: {e}")

    async def close(self):
        print("Disconnecting from Twitch...")
        save_conversations()
        await super().close()


# Function to save conversations
def save_conversations():
    print("Saving conversations...")
    os.makedirs('./conversations/user_profiles/', exist_ok=True)

    # Iterate through chatters and save their conversations
    for username in chatters:
        filename = f"./conversations/user_profiles/conversation_{username}.json"

        if os.path.exists(filename):
            print(f"Found file for {username}: {filename}")
            # You can add further processing for existing files here
        else:
            print(f"No file found for {username}, creating one: {filename}")
            with open(filename, "w") as file:
                json.dump([], file)

    print("Conversations saved.")


# Function to append a message to the JSON file
def append_message(chat_entry):
    os.makedirs('./conversations/chat_logs/', exist_ok=True)
    file_path = './conversations/chat_logs/twitchchatconvervation_peepingotter.json'

    try:
        if not os.path.exists(file_path):
            # Create a new file with an empty list if it doesn't exist
            with open(file_path, "w") as file:
                json.dump([chat_entry], file, indent=2)
        else:
            # Load existing messages and append the new one
            with open(file_path, "r") as file:
                data = json.load(file)

            data.append(chat_entry)

            # If there are more than 80 entries, remove the earliest 40
            if len(data) > 80:
                data = data[40:]

            with open(file_path, "w") as file:
                json.dump(data, file, indent=2)
    except Exception as e:
        print(f"Error appending message: {e}")


# Function to handle bot logic (replace with your implementation)
async def ask_question(message, username):
    # Simulate a response from the bot
    if "hello" in message.lower():
        return "Hello! How can I assist you today?"
    if "nami" in message.lower():
        return "Hi, this is Nami! How can I help you?"
    return None


# Run the bot
bot = TwitchBot()
bot.run()
